### PR TITLE
[X86] Run X86FastPreTileConfigPass only with FastISel.

### LINKED ISSUE #64452: [x86_64/GlobalISel] Crash on `lshr` due to error "Register class not set, wrong accessor" when targeting some CPUs
## Description

When targeting Intel `sapphirerapids` / `icelake-server`, or AMD `znver4` CPU with Global ISel, IR instruction `lshr` crashes the `x86_64` backend with Assertion Error `isa<const TargetRegisterClass *>(VRegInfo[Reg.id()].first) && "Register class not set, wrong accessor"` during `X86 Domain Reassignment Pass`.

The problem cannot be reproduced when targeting `raptorlake`, `alderlake`, `znver3` CPUs, when CPU is not specified, or when Global ISel is not used,

## Minimal Reproduction

https://godbolt.org/z/xY9voTn4z

### Code

```ll
define void @f(i64 %0) {
  %unused = lshr i64 %0, 7
  ret void
}
```
### Stack Trace

```console

llc: /root/llvm-project/llvm/include/llvm/CodeGen/MachineRegisterInfo.h:663: const llvm::TargetRegisterClass* llvm::MachineRegisterInfo::getRegClass(llvm::Register) const: Assertion `isa<const TargetRegisterClass *>(VRegInfo[Reg.id()].first) && "Register class not set, wrong accessor"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/llc -o /app/output.s -x86-asm-syntax=intel -mtriple=x86_64 -mcpu=sapphirerapids --global-isel <source>
1.	Running pass 'Function Pass Manager' on module '<source>'.
2.	Running pass 'X86 Domain Reassignment Pass' on function '@f'
 #0 0x0000000003398418 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x3398418)
 #1 0x0000000003395d0c SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f6293365420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x00007f6292e2800b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #4 0x00007f6292e07859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #5 0x00007f6292e07729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #6 0x00007f6292e18fd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #7 0x0000000001ae99f6 (anonymous namespace)::X86DomainReassignment::runOnMachineFunction(llvm::MachineFunction&) (.part.0) X86DomainReassignment.cpp:0:0
 #8 0x0000000002680049 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
 #9 0x0000000002bf8bb9 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x2bf8bb9)
#10 0x0000000002bf8df1 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x2bf8df1)
#11 0x0000000002bf9612 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x2bf9612)
#12 0x000000000075e092 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#13 0x000000000069a7e6 main (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x69a7e6)
#14 0x00007f6292e09083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#15 0x0000000000754a4e _start (/opt/compiler-explorer/clang-assertions-trunk/bin/llc+0x754a4e)
Program terminated with signal: SIGSEGV
Compiler returned: 139
```

### PR DESCRIPTION
**Pull request by e-kud**
Closes #64452

### CODE REVIEW

**Comment by e-kud** [comment]
This is a revision from the phabricator https://reviews.llvm.org/D157458.
The revision hanged up on a dicussion of whether it is a bug or intended behavior in `GlobalISel` when a register doesn't have a class. I've created an issue for the problem: https://github.com/llvm/llvm-project/issues/66531. Want to bring attention to this issue on the next GlobalISel sync meeting.

It's not `X86` problem. The same can be observed on `AArch64` as well. The only difference is that `X86` is able to reveal this problem. At the same time, `X86FastPreTileConfigPass` was enabled in assumption of `SelectionDAG` and `FastISel` and it crashes with `GlobalISel` if there is a single dead instruction. I want to disable it until it is clear whether `GlobalISel` needs it or not.
Another moment, previous enabling looks suspicious because non-zero opt level doesn't guarantee that `FastISel` is not used.

I intentionally do not fix comments in `getRegClass` as it depends on https://github.com/llvm/llvm-project/issues/66531 resolution. However I can add a comment about the issue somewhere if needed.

/cc @qcolombet @arsenm @aemerson

File: llvm/lib/Target/X86/X86FastPreTileConfig.cpp [hunk 1]
```diff
@@ -667,7 +667,9 @@ bool X86FastPreTileConfig::runOnMachineFunction(MachineFunction &MFunc) {
   bool HasVirtTileReg = false;
   for (unsigned I = 0, E = NumVirtRegs; I != E; ++I) {
     Register VirtReg = Register::index2VirtReg(I);
-    if (MRI->getRegClass(VirtReg)->getID() == X86::TILERegClassID) {
+    const TargetRegisterClass *RC = MRI->getRegClassOrNull(VirtReg);
+    assert(RC && "A register must have a class after FastISel");
```
Comment (arsenm) [inline llvm/lib/Target/X86/X86FastPreTileConfig.cpp:671]: Just use getRegClass instead of doing your own assert

Comment (arsenm) [reply llvm/lib/Target/X86/X86FastPreTileConfig.cpp:671]: We probably should have a registers have a class function property to go with the others that the PM would enforce

**Comment by arsenm** [comment]
This is not the kind of problem that can be correctly solved by changing the pass set. The MIR must follow a consistent set of rules at all points. All selectors must be bound to emit MIR that respects the rules. Shuffling around which passes run with which selector is expected will not work reliably, especially with fallbacks. Every selector should be guaranteeing no virtual registers without classes are passed through

**Comment by e-kud** [comment]
> This is not the kind of problem that can be correctly solved by changing the pass set. The MIR must follow a consistent set of rules at all points. All selectors must be bound to emit MIR that respects the rules. Shuffling around which passes run with which selector is expected will not work reliably, especially with fallbacks. Every selector should be guaranteeing no virtual registers without classes are passed through

Agreed. Also I've realized that generally we want to keep this pass for GlobalISel as it allows a fast skip at O0. Closing in favor of the original issue that leaves registers without classes.

**Pull request closed without merging**